### PR TITLE
Index compact rules for faster filtering

### PR DESCRIPTION
**Pull request by sammacbeth**
Task/Issue URL: https://app.asana.com/1/137249556945/project/488551667048375/task/1211076317769725?focus=true

## Description:
 - Sorts rules in compact format to allow for fast filtering on the client side.
 - Adds an index to allow filtering of generic, frame and specific rules independantly.
 - Adds an implementation of filtering, and enables it in the test extension.
 - Adds tests for the filtering implementation, to ensure that rules are always included when expected.

### CODE REVIEW

File: lib/encoding.ts [hunk 1]
```diff
@@ -43,6 +43,16 @@ export type CompactCMPRuleset = {
     r: CompactCMPRule[];
 };
 
+export type IndexedCMPRuleset = CompactCMPRuleset & {
+    index: {
+        genericRuleRange: [number, number]; // [startIndex, endIndex] of rules that are generic
+        frameRuleRange: [number, number]; // [startIndex, endIndex] of rules that run in frames
+        specificRuleRange: [number, number]; // [startIndex, endIndex] of rules that are specific
+        genericStringEnd: number; // end index of strings that are used by generic rules
+        frameStringEnd: number; // end index of strings that are used by frame rules
```
Comment (muodov) [inline lib/encoding.ts:53]: Could you add a comment about the order inside the strings and rules arrays? So it's easier to understand how these numbers correspond to the actual data structure.

File: lib/encoding.ts [hunk 2]
```diff
@@ -43,6 +43,16 @@ export type CompactCMPRuleset = {
     r: CompactCMPRule[];
 };
 
+export type IndexedCMPRuleset = CompactCMPRuleset & {
```
Comment (muodov) [inline lib/encoding.ts:46]: Is it necessary to keep `CompactCMPRuleset` as a separate type?

File: lib/encoding.ts [hunk 3]
```diff
@@ -131,8 +141,51 @@ export function buildStrings(existingStrings: string[], rules: AutoConsentCMPRul
     return strings;
 }
 
-export function encodeRules(rules: AutoConsentCMPRule[], existingCompactRules: CompactCMPRuleset | null): CompactCMPRuleset {
-    const strings = buildStrings(existingCompactRules?.s || [], rules);
+export function encodeRules(rules: AutoConsentCMPRule[], existingCompactRules: CompactCMPRuleset | null): IndexedCMPRuleset {
+    rules.sort((a, b) => {
+        const isGeneric = (r: AutoConsentCMPRule) => !r.runContext?.urlPattern || r.runContext.urlPattern === '';
```
Comment (muodov) [inline lib/encoding.ts]: this is equivalent, right?
```suggestion
        const isGeneric = (r: AutoConsentCMPRule) => !r.runContext?.urlPattern;
```

File: lib/encoding.ts [hunk 4]
```diff
@@ -131,8 +141,51 @@ export function buildStrings(existingStrings: string[], rules: AutoConsentCMPRul
     return strings;
 }
 
-export function encodeRules(rules: AutoConsentCMPRule[], existingCompactRules: CompactCMPRuleset | null): CompactCMPRuleset {
-    const strings = buildStrings(existingCompactRules?.s || [], rules);
+export function encodeRules(rules: AutoConsentCMPRule[], existingCompactRules: CompactCMPRuleset | null): IndexedCMPRuleset {
```
Comment (muodov) [inline lib/encoding.ts:163]: add a comment/jsdoc about the expected sorting?

File: lib/encoding.ts [hunk 5]
```diff
@@ -131,8 +141,51 @@ export function buildStrings(existingStrings: string[], rules: AutoConsentCMPRul
...
+            if (!isFrame(a) && isFrame(b)) {
+                return 1;
+            }
+        }
+        // all else being equal, sort by name
+        return a.name.localeCompare(b.name);
+    });
+
+    const genericRuleEnd = rules.findIndex((r) => {
+        return r.runContext?.urlPattern && r.runContext.urlPattern !== '';
```
Comment (muodov) [inline lib/encoding.ts]:
```suggestion
        return r.runContext?.urlPattern;
```

File: lib/encoding.ts [hunk 6]
```diff
@@ -131,8 +141,51 @@ export function buildStrings(existingStrings: string[], rules: AutoConsentCMPRul
...
+    });
+    const frameRuleStart = rules.findIndex((r) => {
+        return r.runContext?.frame === true;
+    });
+    const frameRuleEnd = rules.findIndex((r, i) => {
```
Comment (muodov) [inline lib/encoding.ts]: Could you add some tests for edge cases? Such as when there's no frame rules, or no specific rules etc.

File: lib/encoding.ts [hunk 7]
```diff
@@ -131,8 +141,51 @@ export function buildStrings(existingStrings: string[], rules: AutoConsentCMPRul
...
+        return (!r.runContext || !r.runContext?.frame) && i >= frameRuleStart;
```
Comment (muodov) [inline lib/encoding.ts]: this is equivalent, isn't it?
```suggestion
        return !r.runContext?.frame && i >= frameRuleStart;
```

File: lib/encoding.ts [hunk 8]
```diff
@@ -131,8 +141,51 @@ export function buildStrings(existingStrings: string[], rules: AutoConsentCMPRul
...
+    });
+
+    const genericStrings = buildStrings(existingCompactRules?.s || [], rules.slice(0, genericRuleEnd));
+    const frameStrings = buildStrings(genericStrings, rules.slice(0, frameRuleEnd));
+    const strings = buildStrings(frameStrings || [], rules);
```
Comment (muodov) [inline lib/encoding.ts]: Will this (try to) preserve the indices in site-specific rules?

File: lib/encoding.ts (hunk 1 above)
Comment (muodov) [reply lib/encoding.ts:53]: Also, as I understand, all "end" indices are +1 (not inclusive on the right side), right? I'd make an explicit comment about that.

File: lib/encoding.ts [hunk 9]
```diff
@@ -325,3 +385,93 @@ export function deduplicateRules(rules: AutoConsentCMPRule[]) {
...
+            if (idx < ignoreBeforeIndex || usedStringIds.has(idx)) {
+                return str;
+            }
+            return '';
+        }),
+    };
+}
+
+function shouldRunRuleInContext(rule: CompactCMPRule, mainFrame: boolean, url: string): boolean {
+    const runContext = rule[4];
```
Comment (muodov) [inline lib/encoding.ts:462]: Let's change this 4 and the numbers below to constants or enum values? I think it'd be very difficult to understand for anyone new to the codebase.

File: tests-wtr/rules/rule-filtering.test.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+/** @type {IndexedCMPRuleset} */
+import compactRules from '../../rules/compact-rules.json';
+
+describe('filterCompactRules', () => {
+    const decodedRules = decodeRules(compactRules);
+
+    testUrls.forEach(([cmp, testUrl]) => {
+        const rule = decodedRules.find((rule) => rule.name === cmp);
+        if (!rule) {
+            // rule from the test list doesn't exist anymore.
```
Comment (muodov) [inline tests-wtr/rules/rule-filtering.test.ts:13]: Should we regenerate the test urls on each run? Or is it not reliable?

File: lib/encoding.ts (hunk 2 above)
Comment (sammacbeth) [reply lib/encoding.ts:46]: Yes, because post-filtering the index is excluded from the returned result (see the signature of `filterCompactRules`)

File: tests-wtr/rules/rule-filtering.test.ts (hunk 10 above)
Comment (sammacbeth) [reply tests-wtr/rules/rule-filtering.test.ts:13]: Some rules may generated invalid URLs if the regex is not specific enough. I think periodically re-running should be sufficient.

File: lib/encoding.ts (hunk 8 above)
Comment (sammacbeth) [reply lib/encoding.ts]: No, unfortunately not. Do you think it’s worth adding that? It will make this bit of the algorithm a bit messier.

File: lib/encoding.ts (hunk 9 above)
Comment (sammacbeth) [reply lib/encoding.ts:462]: Given that it’s typed, then type is shown. In VSCode you can see the property type on mouseover when using the number index, while using a constant it just shows the constant value. IMO sticking with the raw indices is better DX here:
Screenshot 2025-10-17 at 11 16 47
Screenshot 2025-10-17 at 11 19 02

File: lib/encoding.ts (hunk 1 above)
Comment (sammacbeth) [reply lib/encoding.ts:53]: Added comments this effect.

**Comment by daxtheduck** [comment]
CI run finished. Artifacts [ZIP](https://jenkins.duckduckgo.com/job/cookie%20prompt%20management/job/autoconsent-ci/job/PR-1035/7/artifact/*zip*/archive.zip) for the [review tool](https://zok.pw/autoconsent-review-tool/)

File: lib/encoding.ts (hunk 6 above)
Comment (sammacbeth) [reply lib/encoding.ts]: Added, and fixed some bugs :)

File: lib/encoding.ts (hunk 8 above)
Comment (sammacbeth) [reply lib/encoding.ts]: I found an approach that should work.

**Comment by daxtheduck** [comment]
CI run finished. Artifacts [ZIP](https://jenkins.duckduckgo.com/job/cookie%20prompt%20management/job/autoconsent-ci/job/PR-1035/9/artifact/*zip*/archive.zip) for the [review tool](https://zok.pw/autoconsent-review-tool/)

File: lib/encoding.ts [hunk 11]
```diff
@@ -212,7 +213,8 @@ export function encodeRules(rules: AutoConsentCMPRule[], existingCompactRules: C
 
     const genericStrings = buildStrings(existingCompactRules?.s || [], rules.slice(0, genericRuleEnd));
     const frameStrings = buildStrings(genericStrings, rules.slice(0, frameRuleEnd));
-    const strings = buildStrings(frameStrings, rules);
+    // add back remaining strings from existing ruleset, if any, so specific string ordering can be preserved.
+    const strings = buildStrings(frameStrings.concat(existingCompactRules?.s.slice(frameStrings.length) || []), rules);
```
Comment (muodov) [inline lib/encoding.ts:217]: should we a test for this?

**Comment by daxtheduck** [comment]
CI run finished. Artifacts [ZIP](https://jenkins.duckduckgo.com/job/cookie%20prompt%20management/job/autoconsent-ci/job/PR-1035/13/artifact/*zip*/archive.zip) for the [review tool](https://zok.pw/autoconsent-review-tool/)

**Comment by daxtheduck** [comment]
CI run finished. Artifacts [ZIP](https://jenkins.duckduckgo.com/job/cookie%20prompt%20management/job/autoconsent-ci/job/PR-1035/14/artifact/*zip*/archive.zip) for the [review tool](https://zok.pw/autoconsent-review-tool/)
